### PR TITLE
Added support for generic HTTPException handlers on app and blueprints

### PR DESCRIPTION
**Pull request by cerickson**
Resolves issue #941, replaces pr #2082 and #2144, code based on #2144, tangentially related to #1383
Error handlers are now returned in order of blueprint:code, app:code, blueprint:HTTPException, app:HTTPException, None

Corresponding tests also added.

### CODE REVIEW

**Comment by cerickson** [comment]
This might also close #2197

**Comment by jeffwidman** [comment]
@cerickson I don't know how you feel about this, and it might be beyond your git foo, but what do you think about cherry-picking the initial commit from #2144 as the initial commit in your branch and then making your changes on top of that? The final code would be the same, but that way @sumpfgottheit would get credit for the hard work he did in figuring this out initially and taking the time to write the initial set of tests.

File: flask/app.py [hunk 1]
```diff
@@ -1462,28 +1462,31 @@ def url_defaults(self, f):
         return f
 
     def _find_error_handler(self, e):
-        """Finds a registered error handler for the request’s blueprint.
-        Otherwise falls back to the app, returns None if not a suitable
-        handler is found.
+        """Finds a registered error handlers for a request in this order:
```
Comment (jeffwidman) [inline flask/app.py]: This should be written in the imperative:  `Find` rather than `Finds`

File: flask/app.py [hunk 2]
```diff
@@ -1462,28 +1462,31 @@ def url_defaults(self, f):
...
 
-        # try blueprint handlers
-        handler = find_handler(self.error_handler_spec
-                               .get(request.blueprint, {})
-                               .get(code))
-        if handler is not None:
-            return handler
+        # check for any in blueprint or app
+        for name, c in ((request.blueprint, code), (None, code),
+                        (request.blueprint, None), (None,None)):
```
Comment (jeffwidman) [inline flask/app.py]: The tuple `None,None` is missing a space

**Comment by cerickson** [comment]
I'm not sure how to do that in git, but you are more then welcome to merge the PRs somehow or if you can give me a link or a rough outline of the procedure I'd be happy to try it.  We should absolutely give the original PR credit.

**Comment by jeffwidman** [comment]
Why did you close this?

**Comment by cerickson** [comment]
We don't know, something went weird with git. Can't reopen it so resubmitting it. -davidism

**Pull request closed without merging**